// zcatalyst-integ-cliq/lib/cliq-objects/handler-objects/function-handler.js
import { Handler } from '../../handler.js';
import CONSTANTS from '../../constants.js';
import HandlerResponse from '../response-objects/handler-response.js';
import FormChangeResponse from '../response-objects/form-change-response.js';
import FormDynamicFieldResponse from '../response-objects/form-dynamic-field-response.js';
import WidgetResponse from '../response-objects/widget-response.js';
const { FUNCTION } = CONSTANTS;
const { function_handler } = CONSTANTS.HANDLERS;
export default class FunctionHandler {
    buttonFunctionHandler(handler) {
        Handler.registerHandler(FUNCTION, function_handler.BUTTON_HANDLER, handler, HandlerResponse);
    }
    formSubmitHandler(handler) {
        Handler.registerHandler(FUNCTION, function_handler.FORM_HANDLER, handler, HandlerResponse);
    }
    formChangeHandler(handler) {
        Handler.registerHandler(FUNCTION, function_handler.FORM_CHANGE_HANDLER, handler, FormChangeResponse);
    }
    formDynamicFieldHandler(handler) {
        Handler.registerHandler(FUNCTION, function_handler.FORM_VALUES_HANDLER, handler, FormDynamicFieldResponse);
    }
    widgetButtonHandler(handler) {
        Handler.registerHandler(FUNCTION, function_handler.WIDGET_FUNCTION_HANDLER, handler, WidgetResponse);
    }
    newHandlerResponse() {
        return new HandlerResponse();
    }
    newFormChangeResponse() {
        return new FormChangeResponse();
    }
    newFormDynamicFieldResponse() {
        return new FormDynamicFieldResponse();
    }
    newWidgetResponse() {
        return new WidgetResponse();
    }
}
