// zcatalyst-integ-cliq/lib/cliq-objects/response-objects/handler-response.js
import ButtonObject from './button-object.js';
import CardDetails from './card-details.js';
import Context from './context.js';
import Form from './form.js';
import { Message } from './message.js';
import Slide from './slide.js';
import SuggestionList from './suggestion.js';
export default class HandlerResponse {
    newContext() {
        return new Context();
    }
    newBotDetails(name, image) {
        return {
            name,
            image
        };
    }
    newSuggestionList() {
        return new SuggestionList();
    }
    newSlide() {
        return new Slide();
    }
    addSlide(...slide) {
        if (this.slides === undefined) {
            this.slides = slide;
            return this.slides.length;
        }
        return this.slides.push(...slide);
    }
    newButton() {
        return new ButtonObject();
    }
    addButton(...button) {
        if (this.buttons === undefined) {
            this.buttons = button;
            return this.buttons.length;
        }
        return this.buttons.push(...button);
    }
    newCard() {
        return new CardDetails();
    }
    newForm() {
        return new Form();
    }
    newMessage() {
        return new Message();
    }
    setText(text) {
        this.text = text;
    }
}
