// zcatalyst-integ-cliq/lib/constants.js
import { envOverride } from './util.js';
export default Object.freeze({
    USER_AGENT: 'zcatalyst-integ-node/' + '0.0.1',
    CONFIGJSON: 'catalyst-config.json',
    CLIQ: 'ZohoCliq',
    BOT: 'bot',
    COMMAND: 'command',
    MESSAGEACTION: 'messageaction',
    WIDGET: 'widget',
    FUNCTION: 'function',
    INSTALLATION: 'installation_handler',
    INSTALLATION_VALIDATOR: 'installation_validator',
    HANDLERS: {
        bot_handler: {
            WELCOME_HANDLER: 'welcome_handler',
            MESSAGE_HANDLER: 'message_handler',
            CONTEXT_HANDLER: 'context_handler',
            MENTION_HANDLER: 'mention_handler',
            ACTION_HANDLER: 'action_handler',
            INCOMING_WEBHOOK_HANDLER: 'incomingwebhook_handler',
            PARTICIPATION_HANDLER: 'participation_handler'
        },
        command_handler: {
            EXECUTION_HANDLER: 'execution_handler',
            SUGGESTION_HANDLER: 'suggestion_handler'
        },
        messageaction_handler: {
            EXECUTION_HANDLER: 'execution_handler'
        },
        widget_handler: {
            VIEW_HANDLER: 'view_handler'
        },
        function_handler: {
            BUTTON_HANDLER: 'button_handler',
            FORM_HANDLER: 'form_handler',
            FORM_CHANGE_HANDLER: 'form_change_handler',
            FORM_VALUES_HANDLER: 'form_values_handler',
            WIDGET_FUNCTION_HANDLER: 'applet_button_handler'
        },
        installation_handler: 'installation_handler',
        installation_validator: 'installation_validator'
    },
    API_CONSTANTS: {
        DOMAIN: envOverride('INTEG_CLIQ_DOMAIN', 'https://cliq.zoho.com'),
        VERSION: '/v2',
        API: '/api',
        MESSAGE: '/message',
        FILES: '/files',
        OAUTH_PREFIX: 'Zoho-oauthtoken ',
        URL_CONSTANTS: {
            BOT_UNIQUE_NAME: 'bot_unique_name',
            BROADCAST: 'broadcast',
            USERIDS: 'userids'
        }
    }
});
