// zcatalyst-integ-cliq/lib/cliq-objects/response-objects/form-input.js
import FormValue from './form-value.js';
export default class FormInput {
    newFormValue(label, value) {
        return new FormValue(label, value);
    }
    addOption(...formValue) {
        if (this.options === undefined) {
            this.options = formValue;
            return this.options.length;
        }
        return this.options.push(...formValue);
    }
    /**
     * @deprecated - Misspelled function name
     *
     * Use {@link newBoundary} instead.
     */
    newBoundry(latitude, longitude, radius) {
        // eslint-disable-next-line no-console
        console.warn('Deprecated function - newBoundry');
        return new Boundry(latitude, longitude, radius);
    }
    newBoundary(latitude, longitude, radius) {
        return new Boundary(latitude, longitude, radius);
    }
}
/**
 * @deprecated - class name misspelled
 *
 * Use {@link Boundary} instead.
 */
class Boundry {
    constructor(latitude, longitude, radius) {
        this.latitude = latitude;
        this.longitude = longitude;
        this.radius = radius;
    }
}
class Boundary {
    constructor(latitude, longitude, radius) {
        this.latitude = latitude;
        this.longitude = longitude;
        this.radius = radius;
    }
}
